docs/tpm2-pkcs11_object_auth_model.md

```diff
@@ -0,0 +1,528 @@
+# tpm2-pkcs11 object & authorization model
+
+## Introduction
+
+The objective of this document is to present an architectural overview of the
+tpm2-pkcs11 object and authorization model implemented in the design. The topics
+covered here in are:
+
+1. [PKCS11 Objects](#pkcs11-objects)
+2. [PKCS11 Roles](#pkcs11-roles)
+3. [TPM2 Objects](#tpm2-objects)
+4. [tpm2-pkcs11 Objects](#tpm2-pkcs11-objects)
+5. [NV Object Diagram](#nv-object-diagram)
+6. [NV Policy Diagram](#nv-policy-diagram)
+7. [Transient Object Diagram](#transient-object-diagram)
+8. [Transient Policy Diagram](#transient-policy-diagram)
+9. [Sample tpm2-tools implementation](#sample-tpm2-tools-implementation)
+10. [Important note on authorization](#important-note-on-authorization)
+
+## PKCS11 Objects
+An Overview of the hierarchical concept of objects.
+<p>
+  <img src="illustrations/reader-slot-token-obj.png"
+</p>
+<br>
+<br>
+
+* The PKCS11 READER is a medium through with information is exchanged with the
+PKCS11 device or TOKEN.
+* Every READER has a SLOT that may contain a TOKEN.
+* The TOKEN ultimately protects all the USER created objects as well as access
+through an authorization model to distinguish the SO, USER or adversary.
+* USER objects can be keys, certificates or any user defined data.
+
+## PKCS11 Roles
+Overview of pkcs11 requirements of the security officer and user roles. The
+PKCS11 design describes two specific roles - A security-officer or **SO**
+and **USER**. These roles are realized with separate PIN authentications.
+Following are some relevant design considerations when comparing the roles:
+* Neither SO nor USER can know each other's PIN.
+* The SO is restricted to an administrative role to modify SO PIN and or an
+existing USER's PIN.
+* While both the **SO and the USER can change USER PIN**, only a SO can change
+SO PIN.
+* USER authentication PIN is a must for operating on USER objects.
```

Comment:
USER PIN authentication is a must for TPM operations requiring the authorization be satisfied for the USER objects (Keys) ?

Reply:
Ahh the NIST PIV, a card I know well ([I wrote a clone if it](https://youtu.be/1WG4mJ1HXQw?t=2040)). 

It's quite common that public key/cert objects are made available without a pin, but I think the choice is token specific, like getrandom for example. But it might break existing things.

I'm not sure how the new code drop will work in this regard... but we could always loadexternal the public portion. I cant even recall offhand if I handle this right now. In the new model the primary key is the parent (no more secondary object) so we don't need pin. If you need access to the primary key the pin is either known to the application plaintext (maybe) in the file store or NULL.

I thought about the wrapping key aes/gcm model used to bypass TPM and brute force the pins. We could use non-gcm mode. But most passwords are either blank or reduced to the ascii character set. I figured for the blank case, we just leave the field null, but for the ascii case, for small pins it reduces the difficulty of guessing.

Reply:
> USER PIN authentication is a must for TPM operations requiring the authorization be satisfied for the USER objects (Keys) ?

Imran PKCS11 has User Objects in many flavors, a few being: public, certificate and private. Usually you only need the pin for private key objects.

Reply:
> I'm not sure how the new code drop will work in this regard... but we could always loadexternal the public portion. I cant even recall offhand if I handle this right now. 

We should just use OSSL for the public object portions, no need to incur the round trip penalty. I don't think I tested public key objects yet because all the users where just generating a public key from the components.

I'm still not sure how the new code drop will handle this...